lib/OpenQA/Schema/Result/ScheduledProducts.pm

```diff
@@ -565,22 +569,7 @@ sub _generate_jobs {
 
             # variable expansion
             # replace %NAME% with $settings{NAME}
-            my $expanded;
-            do {
-                $expanded = 0;
-                for my $var (keys %settings) {
-                    my $val = $settings{$var};
-                    next unless ($val && $val =~ /(%\w+%)/);
-                    my $replace_var = $1;
-                    $replace_var =~ s/^%(\w+)%$/$1/;
-                    my $replace_val = $settings{$replace_var};
-                    next unless (defined $replace_val);
-                    $replace_val = '' if $replace_var eq $var;    #stop infinite recursion
-                    $val =~ s/%${replace_var}%/$replace_val/g;
-                    $settings{$var} = $val;
-                    $expanded = 1;
-                }
-            } while ($expanded);
+            $expand_result = OpenQA::ExpandPlaceholder::expand_placeholders(\%settings);
```

Comment:
`$exapnd_result` is `undef` on success and the error message in the error case. I don't see that the loop is aborted in the error case (maybe I'm just overlooking something). So on the next iteration a previously occurred error would be overridden which is likely not what should happen.

Reply:
previously, I returned an error and did not care whether the error message would be changed by following error. As you reminded, I think maybe I should push the error message to a array, when loop is finished, check the array.  Or when error happens, aborted the loop.  which one is better?

Reply:
Adding errors to an array or just concatenate them in a good way would make sense.

Reply:
Fixed it.